### ISSUE TITLE
investigate ways to avoid heap allocations

### ISSUE THREAD

**Issue by mcginty** [mcginty/snow#16]
Right now for the sake of cleaner syntax, there's allocation being done using `Box` structs to avoid forcing users of snow to specify a ton of ugly generics when they store sessions in their own structs, for example.

**Comment by vks** [comment]
It would be an option to use `impl Trait`. Unfortunately, this does not allow users to store the results in their own structs.

Maybe associated types would help?

**Comment by mcginty** [comment]
@vks I think we'd need `impl Trait` to extend support for being used in structs, which it currently can't do and I'm not sure what the roadmap is for that.

**Comment by mcginty** [comment]
related to #62

**Comment by isislovecruft** [comment]
Why is `snow` doing `Box<dyn SomeTrait>` everywhere rather than `&'a SomeTrait`? Is it because the explicit lifetimes would make it harder for `snow` users?

**Comment by mcginty** [comment]
@isislovecruft yep, it's purely for ergonomic reasons. I'd love to use something like associated traits and maybe a convenience macro to be able to drop the `Box`es while keeping roughly the same level of simplicity.

**Comment by tarcieri** [comment]
If you want trait objects without heap allocations, one option is an `enum` with `Deref` impl whose `Target` is the desired trait object. Something like this:

```rust
enum CipherBackends {
    #[cfg(feature = "backend1")]
    ....
}
```

```rust
impl core::ops::Deref for CipherBackends {
    type Target = dyn Cipher;

    fn deref(&self) -> &(dyn Cipher + 'static) {
        match self {
            #[cfg(feature = "backend1")]
            ....
        }
    }
}
```

Note that this does have the advantage that the size of `CipherBackends` will always be as big as the largest of the variants, but hopefully they're roughly the same size.

**Comment by mcginty** [comment]
@tarcieri the reason I didn't choose this method in the original design is because I wanted the option for people to implement their own crypto backends if snow didn't provide any that fit their needs. If it's totally clear that flexibility isn't necessary, we could definitely switch to something like this.

**Comment by tarcieri** [comment]
Unfortunately I think you need either `Box` or a generic type parameter in that case, the former requiring a heap allocation, and the latter defeating the point of using an `enum` since it can just monomorphize rather than using a trait object.
